processor/partition_worker.go

```diff
@@ -0,0 +1,135 @@
+package processor
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/services/rsources"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+type partitionWorker struct {
+	partition string
+	workers   []*pipelineWorker
+	logger    logger.Logger
+	stats     *processorStats
+	handle    workerHandle
+	g         *errgroup.Group
+	ctx       context.Context
+}
+
+// newProcessorWorker creates a new worker for the specified partition
+func newPartitionWorker(ctx context.Context, partition string, h workerHandle) *partitionWorker {
+	w := &partitionWorker{
+		partition: partition,
+		logger:    h.logger().Child(partition),
+		stats:     h.stats(),
+		handle:    h,
+	}
+	w.g, w.ctx = errgroup.WithContext(ctx)
+
+	// Create workers for each partition
+	pipelinesPerPartition := h.config().pipelinesPerPartition
+	w.workers = make([]*pipelineWorker, pipelinesPerPartition)
+	for i := 0; i < pipelinesPerPartition; i++ {
+		w.workers[i] = newPipelineWorker(partition, h)
+	}
+
+	return w
+}
+
+// Work processes jobs for the specified partition
+// Returns true if work was done, false otherwise
+func (w *partitionWorker) Work() bool {
+	// If pipelining is disabled, use the legacy job handling path
+	if !w.handle.config().enablePipelining {
+		return w.handle.handlePendingGatewayJobs(w.partition)
+	}
+
+	start := time.Now()
+	// Get jobs for this partition
+	jobs := w.handle.getJobs(w.partition)
+
+	// If no jobs were found, return false
+	if len(jobs.Jobs) == 0 {
+		return false
+	}
+
+	// Mark jobs as executing
+	if err := w.handle.markExecuting(w.partition, jobs.Jobs); err != nil {
+		w.logger.Error("Error marking jobs as executing", "error", err)
+		panic(err)
+	}
+
+	// Record throughput metrics
+	w.handle.stats().DBReadThroughput(w.partition).Count(throughputPerSecond(jobs.EventsCount, time.Since(start)))
```

Comment:
I cannot find us using this metric (`processor_db_read_throughput`) in any of our dashboards or alerts. I propose to delete it. We capture throughput through jobsdb metrics now